### ISSUE TITLE
Table schema sync performance impact

### ISSUE THREAD

**Issue by KEN-LJQ** [metabase/metabase#13650]
**Describe the bug**
We use presto over hive as our warehouse, after upgrade to 0.37.x, we found that the new improvement #10892 ,whether table have select privilege,  fire a lot of query like `SELECT * FROM XXX  LIMIT 1` on each table. Finally cause hive metastore overload and return read timeout, which make table sync fail and serious performance hit on common query.
I wonder is there any better approach to avoid this problem?

**Logs**
```
[{:type "org.apache.thrift.transport.TTransportException",
      :message "xxxx:9083: java.net.SocketTimeoutException: Read timed out",
      :cause
      {:type "org.apache.thrift.transport.TTransportException",
       :message "java.net.SocketTimeoutException: Read timed out",
       :cause
       {:type "java.net.SocketTimeoutException",
        :message "Read timed out",
        :suppressed [],
        :stack
        ["java.net.SocketInputStream.socketRead0(Native Method)"
         "java.net.SocketInputStream.socketRead(SocketInputStream.java:116)"
         "java.net.SocketInputStream.read(SocketInputStream.java:171)"
         "java.net.SocketInputStream.read(SocketInputStream.java:141)"
         "java.io.BufferedInputStream.fill(BufferedInputStream.java:246)"
         "java.io.BufferedInputStream.read1(BufferedInputStream.java:286)"
         "java.io.BufferedInputStream.read(BufferedInputStream.java:345)"
         "org.apache.thrift.transport.TIOStreamTransport.read(TIOStreamTransport.java:127)"
         "org.apache.thrift.transport.TTransport.readAll(TTransport.java:84)"
         "com.facebook.presto.hive.metastore.thrift.Transport$TTransportWrapper.readAll(Transport.java:165)"
         "org.apache.thrift.protocol.TBinaryProtocol.readAll(TBinaryProtocol.java:362)"
         "org.apache.thrift.protocol.TBinaryProtocol.readI32(TBinaryProtocol.java:284)"
         "org.apache.thrift.protocol.TBinaryProtocol.readMessageBegin(TBinaryProtocol.java:191)"
         "org.apache.thrift.TServiceClient.receiveBase(TServiceClient.java:69)"
         "org.apache.hadoop.hive.metastore.api.ThriftHiveMetastore$Client.recv_get_table(ThriftHiveMetastore.java:1218)"
         "org.apache.hadoop.hive.metastore.api.ThriftHiveMetastore$Client.get_table(ThriftHiveMetastore.java:1204)"
         "com.facebook.presto.hive.metastore.thrift.ThriftHiveMetastoreClient.getTable(ThriftHiveMetastoreClient.java:148)"
         "com.facebook.presto.hive.metastore.thrift.ThriftHiveMetastore.lambda$getTable$5(ThriftHiveMetastore.java:229)"
         "com.facebook.presto.hive.metastore.thrift.HiveMetastoreApiStats.lambda$wrap$0(HiveMetastoreApiStats.java:42)"
         "com.facebook.presto.hive.RetryDriver.run(RetryDriver.java:139)"
         "com.facebook.presto.hive.metastore.thrift.ThriftHiveMetastore.getTable(ThriftHiveMetastore.java:227)"
         "com.facebook.presto.hive.metastore.thrift.BridgingHiveMetastore.getTable(BridgingHiveMetastore.java:83)"
         "com.facebook.presto.hive.metastore.CachingHiveMetastore.loadTable(CachingHiveMetastore.java:275)"
         "com.google.common.cache.CacheLoader$FunctionToCacheLoader.load(CacheLoader.java:165)"
         "com.google.common.cache.CacheLoader$1.load(CacheLoader.java:188)"
         "com.google.common.cache.LocalCache$LoadingValueReference.loadFuture(LocalCache.java:3524)"
         "com.google.common.cache.LocalCache$Segment.loadSync(LocalCache.java:2273)"
         "com.google.common.cache.LocalCache$Segment.lockedGetOrLoad(LocalCache.java:2156)"
         "com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2046)"
         "com.google.common.cache.LocalCache.get(LocalCache.java:3943)"
         "com.google.common.cache.LocalCache.getOrLoad(LocalCache.java:3967)"
         "com.google.common.cache.LocalCache$LocalLoadingCache.get(LocalCache.java:4952)"
         "com.google.common.cache.LocalCache$LocalLoadingCache.getUnchecked(LocalCache.java:4958)"
         "com.facebook.presto.hive.metastore.CachingHiveMetastore.get(CachingHiveMetastore.java:219)"
         "com.facebook.presto.hive.metastore.CachingHiveMetastore.getTable(CachingHiveMetastore.java:264)"
         "com.facebook.presto.hive.metastore.CachingHiveMetastore.loadTable(CachingHiveMetastore.java:275)"
         "com.google.common.cache.CacheLoader$FunctionToCacheLoader.load(CacheLoader.java:165)"
         "com.google.common.cache.CacheLoader$1.load(CacheLoader.java:188)"
         "com.google.common.cache.LocalCache$LoadingValueReference.loadFuture(LocalCache.java:3524)"
         "com.google.common.cache.LocalCache$Segment.loadSync(LocalCache.java:2273)"
         "com.google.common.cache.LocalCache$Segment.lockedGetOrLoad(LocalCache.java:2156)"
         "com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2046)"
         "com.google.common.cache.LocalCache.get(LocalCache.java:3943)"
         "com.google.common.cache.LocalCache.getOrLoad(LocalCache.java:3967)"
         "com.google.common.cache.LocalCache$LocalLoadingCache.get(LocalCache.java:4952)"
         "com.google.common.cache.LocalCache$LocalLoadingCache.getUnchecked(LocalCache.java:4958)"
         "com.facebook.presto.hive.metastore.CachingHiveMetastore.get(CachingHiveMetastore.java:219)"
         "com.facebook.presto.hive.metastore.CachingHiveMetastore.getTable(CachingHiveMetastore.java:264)"
         "com.facebook.presto.hive.metastore.SemiTransactionalHiveMetastore.getTable(SemiTransactionalHiveMetastore.java:163)"
         "com.facebook.presto.hive.HiveMetadata.getViews(HiveMetadata.java:1678)"
         "com.facebook.presto.spi.connector.classloader.ClassLoaderSafeConnectorMetadata.getViews(ClassLoaderSafeConnectorMetadata.java:438)"
         "com.facebook.presto.metadata.MetadataManager.getView(MetadataManager.java:945)"
         "com.facebook.presto.sql.analyzer.StatementAnalyzer$Visitor.visitTable(StatementAnalyzer.java:870)"
         "com.facebook.presto.sql.analyzer.StatementAnalyzer$Visitor.visitTable(StatementAnalyzer.java:257)"
         "com.facebook.presto.sql.tree.Table.accept(Table.java:53)"
         "com.facebook.presto.sql.tree.AstVisitor.process(AstVisitor.java:27)"
         "com.facebook.presto.sql.analyzer.StatementAnalyzer$Visitor.process(StatementAnalyzer.java:271)"
         "com.facebook.presto.sql.analyzer.StatementAnalyzer$Visitor.analyzeFrom(StatementAnalyzer.java:1852)"
         "com.facebook.presto.sql.analyzer.StatementAnalyzer$Visitor.visitQuerySpecification(StatementAnalyzer.java:1031)"
         "com.facebook.presto.sql.analyzer.StatementAnalyzer$Visitor.visitQuerySpecification(StatementAnalyzer.java:257)"
         "com.facebook.presto.sql.tree.QuerySpecification.accept(QuerySpecification.java:127)"
         "com.facebook.presto.sql.tree.AstVisitor.process(AstVisitor.java:27)"
         "com.facebook.presto.sql.analyzer.StatementAnalyzer$Visitor.process(StatementAnalyzer.java:271)"
         "com.facebook.presto.sql.analyzer.StatementAnalyzer$Visitor.process(StatementAnalyzer.java:281)"
         "com.facebook.presto.sql.analyzer.StatementAnalyzer$Visitor.visitQuery(StatementAnalyzer.java:746)"
         "com.facebook.presto.sql.analyzer.StatementAnalyzer$Visitor.visitQuery(StatementAnalyzer.java:257)"
         "com.facebook.presto.sql.tree.Query.accept(Query.java:94)"
         "com.facebook.presto.sql.tree.AstVisitor.process(AstVisitor.java:27)"
         "com.facebook.presto.sql.analyzer.StatementAnalyzer$Visitor.process(StatementAnalyzer.java:271)"
         "com.facebook.presto.sql.analyzer.StatementAnalyzer.analyze(StatementAnalyzer.java:249)"
         "com.facebook.presto.sql.analyzer.StatementAnalyzer$Visitor.visitTableSubquery(StatementAnalyzer.java:1021)"
         "com.facebook.presto.sql.analyzer.StatementAnalyzer$Visitor.visitTableSubquery(StatementAnalyzer.java:257)"
         "com.facebook.presto.sql.tree.TableSubquery.accept(TableSubquery.java:53)"
         "com.facebook.presto.sql.tree.AstVisitor.process(AstVisitor.java:27)"
         "com.facebook.presto.sql.analyzer.StatementAnalyzer$Visitor.process(StatementAnalyzer.java:271)"
         "com.facebook.presto.sql.analyzer.StatementAnalyzer$Visitor.visitAliasedRelation(StatementAnalyzer.java:954)"
         "com.facebook.presto.sql.analyzer.StatementAnalyzer$Visitor.visitAliasedRelation(StatementAnalyzer.java:257)"
         "com.facebook.presto.sql.tree.AliasedRelation.accept(AliasedRelation.java:71)"
         "com.facebook.presto.sql.tree.AstVisitor.process(AstVisitor.java:27)"
         "com.facebook.presto.sql.analyzer.StatementAnalyzer$Visitor.process(StatementAnalyzer.java:271)"
         "com.facebook.presto.sql.analyzer.StatementAnalyzer$Visitor.visitJoin(StatementAnalyzer.java:1184)"
         "com.facebook.presto.sql.analyzer.StatementAnalyzer$Visitor.visitJoin(StatementAnalyzer.java:257)"
         "com.facebook.presto.sql.tree.Join.accept(Join.java:90)"
         "com.facebook.presto.sql.tree.AstVisitor.process(AstVisitor.java:27)"
         "com.facebook.presto.sql.analyzer.StatementAnalyzer$Visitor.process(StatementAnalyzer.java:271)"
         "com.facebook.presto.sql.analyzer.StatementAnalyzer$Visitor.analyzeFrom(StatementAnalyzer.java:1852)"
         "com.facebook.presto.sql.analyzer.StatementAnalyzer$Visitor.visitQuerySpecification(StatementAnalyzer.java:1031)"
         "com.facebook.presto.sql.analyzer.StatementAnalyzer$Visitor.visitQuerySpecification(StatementAnalyzer.java:257)"
         "com.facebook.presto.sql.tree.QuerySpecification.accept(QuerySpecification.java:127)"
         "com.facebook.presto.sql.tree.AstVisitor.process(AstVisitor.java:27)"
         "com.facebook.presto.sql.analyzer.StatementAnalyzer$Visitor.process(StatementAnalyzer.java:271)"
         "com.facebook.presto.sql.analyzer.StatementAnalyzer$Visitor.process(StatementAnalyzer.java:281)"
         "com.facebook.presto.sql.analyzer.StatementAnalyzer$Visitor.visitQuery(StatementAnalyzer.java:746)"
         "com.facebook.presto.sql.analyzer.StatementAnalyzer$Visitor.visitQuery(StatementAnalyzer.java:257)"
         "com.facebook.presto.sql.tree.Query.accept(Query.java:94)"
         "com.facebook.presto.sql.tree.AstVisitor.process(AstVisitor.java:27)"
         "com.facebook.presto.sql.analyzer.StatementAnalyzer$Visitor.process(StatementAnalyzer.java:271)"
         "com.facebook.presto.sql.analyzer.StatementAnalyzer$Visitor.process(StatementAnalyzer.java:281)"
         "com.facebook.presto.sql.analyzer.StatementAnalyzer$Visitor.analyzeWith(StatementAnalyzer.java:2022)"
         "com.facebook.presto.sql.analyzer.StatementAnalyzer$Visitor.visitQuery(StatementAnalyzer.java:745)"
         "com.facebook.presto.sql.analyzer.StatementAnalyzer$Visitor.visitQuery(StatementAnalyzer.java:257)"
         "com.facebook.presto.sql.tree.Query.accept(Query.java:94)"
         "com.facebook.presto.sql.tree.AstVisitor.process(AstVisitor.java:27)"
         "com.facebook.presto.sql.analyzer.StatementAnalyzer$Visitor.process(StatementAnalyzer.java:271)"
         "com.facebook.presto.sql.analyzer.StatementAnalyzer.analyze(StatementAnalyzer.java:249)"
         "com.facebook.presto.sql.analyzer.Analyzer.analyze(Analyzer.java:79)"
         "com.facebook.presto.sql.analyzer.Analyzer.analyze(Analyzer.java:71)"
         "com.facebook.presto.execution.SqlQueryExecution.<init>(SqlQueryExecution.java:209)"
         "com.facebook.presto.execution.SqlQueryExecution.<init>(SqlQueryExecution.java:97)"
         "com.facebook.presto.execution.SqlQueryExecution$SqlQueryExecutionFactory.createQueryExecution(SqlQueryExecution.java:744)"
         "com.facebook.presto.execution.SqlQueryManager.createQueryInternal(SqlQueryManager.java:366)"
         "com.facebook.presto.execution.SqlQueryManager.lambda$createQuery$4(SqlQueryManager.java:305)"
         "com.facebook.presto.$gen.Presto_0_220____20201009_011753_1.run(Unknown Source)"
         "java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)"
         "java.util.concurrent.FutureTask.run(FutureTask.java:266)"
         "java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)"
         "java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)"
         "java.lang.Thread.run(Thread.java:748)"]}
```

**Information about your Metabase Installation:**

```
{
  "browser-info": {
    "language": "zh-CN",
    "platform": "Win32",
    "userAgent": "Mozilla/5.0 (Windows NT 6.1; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/86.0.4240.111 Safari/537.36",
    "vendor": "Google Inc."
  },
  "system-info": {
    "file.encoding": "UTF-8",
    "java.runtime.name": "Java(TM) SE Runtime Environment",
    "java.runtime.version": "1.8.0_162-b12",
    "java.vendor": "Oracle Corporation",
    "java.vendor.url": "http://java.oracle.com/",
    "java.version": "1.8.0_162",
    "java.vm.name": "Java HotSpot(TM) 64-Bit Server VM",
    "java.vm.version": "25.162-b12",
    "os.name": "Linux",
    "os.version": "2.6.32-696.3.1.el6.x86_64",
    "user.language": "zh",
    "user.timezone": "Asia/Shanghai"
  },
  "metabase-info": {
    "databases": [
      "h2",
      "presto"
    ],
    "hosting-env": "unknown",
    "application-database": "mysql",
    "application-database-details": {
      "database": {
        "name": "MySQL",
        "version": "5.7.26-log"
      },
      "jdbc-driver": {
        "name": "MariaDB Connector/J",
        "version": "2.6.2"
      }
    },
    "run-mode": "prod",
    "version": {
      "tag": "v0.37.0.1",
      "date": "",
      "branch": "2020-10-28",
      "hash": "pre-production"
    },
    "settings": {
      "report-timezone": "Asia/Hong_Kong"
    }
  }
}
```

**Severity**
Apparently query with limit 1 has difference hehavior within difference database, in our case it hit a lot of metastore query. Our temporary solution is improve hive metatore performance, modify and fire a query with where clause ` 1 <> 1` to avoid partition scan.
I don't know if we doing thing right, hopeful for your reply.

**Comment by flamber** [comment]
Changing priority based on https://discourse.metabase.com/t/metadata-lightweight-sync-taking-too-long/12204:
> ...
> He’s also noticed a degradation in our database’s performance because of this due to how long the queries were stuck
> ...
> The sync finished but I see it took a very long time: [9417aef0-a561-4977-8594-8741751cfdef] 2020-11-03T14:03:38-05:00 INFO metabase.sync.util FINISHED: Sync metadata for sqlserver Database 34 ‘Database_Name’ (17.1 hours)

**Comment by salsakran** [comment]
@KEN-LJQ perhaps a dumb question, but does ` SELECT FROM {{TableName}} WHERE 1 <> 1` fail on tables you don't have read access to on presto?

**Comment by KEN-LJQ** [comment]
@salsakran yes , we’ve been test it and it work fine. We want to reduce the overhead of this operation but don’t loss the select privileges test feature.

**Comment by camsaul** [comment]
Fixed by #13746
